netman/dbus.go

```diff
@@ -135,104 +171,111 @@ func getSystemBus() *dbus.Conn {
-		return false
-	}
-	if nmConnectivityStatus == 1 {
-		return false //not connected by any netman interfaces
+// check if no ethernet or wifi devices are connected
+func (c *Client) Connected(devices []string) bool {
+	for _, d := range devices {
+		objPath := dbus.ObjectPath(d)
+		c.dbusClient.Object("org.freedesktop.NetworkManager", objPath)
+		setObject(c, "org.freedesktop.NetworkManager", objPath)
+		dType, _ := c.dbusClient.BusObj.GetProperty("org.freedesktop.NetworkManager.Device.DeviceType")
+		state, _ := c.dbusClient.BusObj.GetProperty("org.freedesktop.NetworkManager.Device.State")
+		if dbus.Variant.Value(dType) != uint32(1) && dbus.Variant.Value(dType) != uint32(2) {
+			continue
+		}
+		if dbus.Variant.Value(state) == uint32(100) {
+			return true
+		}
 	}
-	return true
+	return false
 }
 
-// DisconnectWifi disconnect current network
-func DisconnectWifi() {
-	conn := getSystemBus()
-	devices := getDevices()
-	wifiDevices := getWifiDevices(devices)
+// Check if connected by wifi
+func (c *Client) ConnectedWifi(wifiDevices []string) bool {
 	for _, d := range wifiDevices {
 		objPath := dbus.ObjectPath(d)
-		device := conn.Object("org.freedesktop.NetworkManager", objPath)
-		device.Call("org.freedesktop.NetworkManager.Device.Disconnect", 0)
+		c.dbusClient.Object("org.freedesktop.NetworkManager", objPath)
+		setObject(c, "org.freedesktop.NetworkManager", objPath)
+		state, _ := c.dbusClient.BusObj.GetProperty("org.freedesktop.NetworkManager.Device.State")
+		if dbus.Variant.Value(state) == uint32(100) {
+			return true
+		}
 	}
-	return
+	return false
 }
 
-// SetIfaceManaged sets certain interface managed by netman
-func SetIfaceManaged(iface string) {
-	conn := getSystemBus()
-	devices := getDevices()
+// Disconnects every interface passed. return shows number of disconnect calls  made
+func (c *Client) DisconnectWifi(wifiDevices []string) int {
+	ran := 0
+	for _, d := range wifiDevices {
+		objPath := dbus.ObjectPath(d)
+		c.dbusClient.Object("org.freedesktop.NetworkManager", objPath)
+		setObject(c, "org.freedesktop.NetworkManager", objPath)
+		c.dbusClient.BusObj.Call("org.freedesktop.NetworkManager.Device.Disconnect", 0)
+		ran += 1
+	}
+	return ran
+}
 
+// Set passed device to be managed by network manager, return iface set, if any
+func (c *Client) SetIfaceManaged(iface string, devices []string) string {
+	ran := ""
 	for _, d := range devices {
 		objPath := dbus.ObjectPath(d)
-		device := conn.Object("org.freedesktop.NetworkManager", objPath)
-		deviceIface, err2 := device.GetProperty("org.freedesktop.NetworkManager.Device.Interface")
+		c.dbusClient.Object("org.freedesktop.NetworkManager", objPath)
+		setObject(c, "org.freedesktop.NetworkManager", objPath)
+		iface_, err2 := c.dbusClient.BusObj.GetProperty("org.freedesktop.NetworkManager.Device.Interface")
 		if err2 != nil {
-			fmt.Printf("Error in SetIfaceManaged(): %v\n", err2)
-			return
+			fmt.Printf("Error 1 in SetIfaceManaged(): %v\n", err2)
```

Comment:
Instead of Error 1 or Error 2, maybe we could have a more specific message like "Error when getting device interface property" or similar